text/0000-custom-dependency-groups.md

```diff
@@ -0,0 +1,288 @@
+* https://www.npmjs.com/package/dependency-groups
+ 
+
+# Detailed design
+
+My proposal is to:
+
+1/ expand the `dependencies` entry in a `package.json` file to allow organizing the dependencies into user-defined (with "user" being the developer in this case) groups.
+
+A valid `dependencies` entry would either have the current, existing format:
+
+```json
+{
+  "dependencies":{
+    "<dependency name>": "<version range || url>"
+  }
+}
+```
+
+**Or** the newly proposed:
+
+```json
+{
+  "dependencies": {
+    "<group name>": {
+      "<dependency name>": "<version range || url>"
+    }
+  }
+}
+```
+
+No nesting would be allowed (groups in groups in groups).
+
+2/ Further more I'd like to propose a single pre-defined "magic" group name: `default`:
+
+```json
+{
+  "dependencies": {
+    "<group name>": {
+      "<dependency name>": "<version range || url>"
+    },
+    "default": {
+      "<dependency name>": "<version range || url>"
+    }
+  }
+}
+```
+
+The `default` dependencies would always get installed, see below.
+
+3/ Installation would allow passing these group names as switches to the command-line client, denoted with double dashes (`--`):
+
+```sh
+$ yarn --<group name> --<group name> ...
+```
+
+None, one or multiple group names could be passed, making all of these valid:
+
+```sh
+# install only the "default" dependencies
+$ yarn
+
+# install the "default" dependencies and those grouped as "a"
+$ yarn --a
+
+# install the "default" dependencies and those grouped as "a", "b" and "c"
+$ yarn --a --b --c
+```
+
+In case a `default` entry is not declared and no group names are passed to the command-line client an appropriate warning should be thrown.
+
+4/ Adding a dependency would use the same switches as above:
+
+```
+# add package "foo" to the "a" dependency group
+$ yarn add foo --a
+```
+
+However in this case only one switch would be allowed.
+
+```
+# WARNING would be thrown
+$ yarn add foo --a --b
+```
+
+5/ It would NOT be allowed to mix the proposed group dependency declarations with the old style pre-defined top-level group types:
+
+```json
+// INVALID
+{
+  "dependencies": {
+    "<group name>": {
+      "<dependency name>": "<version range || url>"
+    }
+  },
+  "devDependencies": {
+    "<dependency name>": "<version range || url>"
+  }
+}
+```
+
+I.e. either you stick with the npm-compatible dependency declarations -or- you use the new style.
+
+
+# How We Teach This
+
+Documentation and education wise this would be an addition, in the sense that all existing documentation would still be valid. 
+
+Anybody familiar with "dependencies" and "devDependencies" would pick this up pretty fast, I think.
+
+# Drawbacks
+
+The major drawback would be that `package.json` files utilizing the new dependencies declaration style would no longer be recognized as valid manifest files by the `npm` CLI. I.e. once you go down this road, you can't use npm for dependency management in your project anymore.
+
```

Comment:
Additional drawbacks:

- You never know what dependencies are actually needed for what. If you run `./scripts/do-something.js` while inside the repo, how do you know if you have the correct dependencies installed. You don't.
- You would still have to resolve the entire tree from the registry in order to be deterministic, this doesn't improve that at all.
- You complicate the install process so that every single project is different. You move from `git clone repo && yarn install` to `git clone repo && less package.json && yarn install --some --groups --that --hopefully --you --want`. Try something, doesn't work? Try again with the dependencies to install.
- You can't make it so that one category depends on another. ex. "test" needs "app", but there's nothing to say that)
- Dependencies don't fall neatly into any one category. ex. How do you say something is a "test" dependency vs a "build" dependency if it's being used for both? The line is never clear.